core/src/main/java/org/apache/brooklyn/core/mgmt/internal/LocalEntityManager.java

```diff
@@ -212,13 +213,41 @@ public EntityTypeRegistry getEntityTypeRegistry() {
 
     @Override
     public Iterable<Entity> getAllEntitiesInApplication(Application application) {
+        // To fix https://issues.apache.org/jira/browse/BROOKLYN-352, we need to synchronize on
+        // preRegisteredEntitiesById and preManagedEntitiesById while iterating over them (because
+        // they are synchronizedMaps). entityProxiesById is a ConcurrentMap, so no need to 
+        // synchronize on that.
+        // Only synchronize on one at a time, to avoid the risk of deadlock.
+        
         Predicate<Entity> predicate = EntityPredicates.applicationIdEqualTo(application.getId());
-        Iterable<Entity> allentities = Iterables.concat(preRegisteredEntitiesById.values(), preManagedEntitiesById.values(), entityProxiesById.values());
-        Iterable<Entity> result = Iterables.filter(allentities, predicate);
-        return ImmutableSet.copyOf(Iterables.transform(result, new Function<Entity, Entity>() {
-            @Override public Entity apply(Entity input) {
-                return Entities.proxy(input);
-            }}));
+        Set<Entity> result = Sets.newLinkedHashSet();
+        
+        synchronized (preRegisteredEntitiesById) {
```

Comment:
Maybe it would be worth making `preRegisteredEntitiesById` and `preManagedEntitiesById` private, and providing access to them with code like this loop, but put into a `protected` method.  At the moment it's not impossible that some derived class in future would again use one of the maps but forget the synchronization.